### ISSUE TITLE
Netlink: Cannot open netlink socket : (Too many open files)

### ISSUE THREAD

**Issue by TwitchChen** [acassen/keepalived#1331]
**keepalived version**
1.2.13

**problem**
I got a error in the keepalived.log as the tittle. How and why, Does this affect VIP mounts?
when i use the lsof，i got this
```
keepalive 4634 root 1018u  netlink                         0t0 3138431695 ROUTE
keepalive 4634 root 1019u  netlink                         0t0 3138446513 ROUTE
keepalive 4634 root 1020u  netlink                         0t0 3138431731 ROUTE
keepalive 4634 root 1021u  netlink                         0t0 3138451554 ROUTE
keepalive 4634 root 1022u  netlink                         0t0 3138423666 ROUTE
keepalive 4634 root 1023u  netlink                         0t0 3138419698 ROUTE
[root@cnsz081823 ~]# lsof -p 4634 |wc -l
1047
```
ps: I foud that, when i reload keepalived a time ,the ROUTE fd will add 1

the limit
```
[root@cnsz081823 ~]# cat /proc/4634/limits 
Limit                     Soft Limit           Hard Limit           Units     
Max cpu time              unlimited            unlimited            seconds   
Max file size             unlimited            unlimited            bytes     
Max data size             unlimited            unlimited            bytes     
Max stack size            8388608              unlimited            bytes     
Max core file size        0                    unlimited            bytes     
Max resident set          unlimited            unlimited            bytes     
Max processes             255529               255529               processes 
Max open files            1024                 4096                 files     
Max locked memory         65536                65536                bytes     
Max address space         unlimited            unlimited            bytes     
Max file locks            unlimited            unlimited            locks     
Max pending signals       255529               255529               signals   
Max msgqueue size         819200               819200               bytes     
Max nice priority         0                    0                    
Max realtime priority     0                    0                    
Max realtime timeout      unlimited            unlimited            us        
```

**Comment by pqarmitage** [comment]
As can be seen from the lsof output, keepalived has file descriptors open up to 1023, so counting from 0 that is 1024 file descriptors, and as the limits outputs shows that the limit of the number of open files, hence opening another one is failing. And yes, that will cause keepalived not to behave properly.

As you have identified, keepalived is leaking a file descriptor on reload.

We can't investigate this any further, since you have not provided the information requested in the template for reporting a bug, but in any event, v1.2.13 is far too old. Please upgrade to a recent version of keepalived (v2.0.16 is the current version) and see if that resolves your problem.

**Comment by TwitchChen** [comment]
@pqarmitage Thank you for your reply.
when i reload keepalived with or without vip changes, the file descriptors will add one and And will not be reducing,Is this normal for this？

And as seen in the proc limit file , the Hard Limit   of Max open files is 4096, why the file descriptors open up to 1023 and the “too many files open” error occurred

**Comment by pqarmitage** [comment]
@TwitchChen Can you please provide the information requested in the bug report template.

keepalived shouldn't leak a file descriptor on reload, but without the further information requested, including your configuration, OS version etc, etc, we cannot attempt to reproduce the problem or investigate further. The Bug Report issue template sets out the information needed, and should be used for reporting such issues.

Version 1.2.13 is extremely old, and it is highly likely that this issue has been resolved somewhere between v1.2.13 and v2.0.16 (the current version).

**Comment by TwitchChen** [comment]
@pqarmitage sorry for the wrong issue format.

**Describe the bug**
When reloading the keepalived(version 1.2.13) with or without vip changes, the file descriptors will add one and And will not be reducing
 When the Hard Limit of Max open files is 4096 in the proc limit file,the file descriptors open up to 1023 biggest and the “too many files open” error occurred

**To Reproduce**
as the bug description

**Expected behavior**
when there is no vip changes, the file descriptors will be no change or the extra socket shoud be closed
when i set the limitopfile=4096， the error of "too many files open" shoudn'd be occurred when fd is just up to 1023

**Keepalived version**
Keepalived v1.2.13

**Distro (please complete the following information):**
 - Name centos
 - Version centos ( kernel 3.10)
 - Architecture x86_64

**Details of any containerisation or hosted service (e.g. AWS)**
run as a systemd service on the host

**Configuration file:**
there are 3 keepalived have the same conf file
```
global_defs {                                
   router_id keepalived                      
   vrrp_mcast_group4 224.0.0.118             
}
vrrp_script chk_100 {
    script "/etc/keepalived/scripts/check.sh"
    interval 10
    timeout 10
}
vrrp_instance redis_100 {                  
        state BACKUP                         
        interface team0                      
        virtual_router_id 100                 
        priority 120                         
        preempt
        advert_int 1                         
        authentication {                     
            auth_type PASS                   
            auth_pass 1111                   
        }                                    
        virtual_ipaddress { 
            10.25.88.149
        }                                    
        track_script {                       
            chk_100                        
        }                                    
}
.....
there are another 100 chk_*、redis_* like this in the file
```

**Notify and track scripts**
```
#!/bin/bash
LOGFILE=/pacloud/cdm/base/1.0/logs/check.log  
echo "[check]" >> $LOGFILE
date >> $LOGFILE
#check nginx 
cnt1=`ps -ef|grep nginx|grep -v grep |wc -l `
if [ $cnt1 -gt 0 ] ; then 
   echo "[check] nginx is ok " >>$LOGFILE
   exit 0
else
   echo "[check] nginx is stop. " >>$LOGFILE 
   exit 1
fi
```
**System Log entries**
```
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Got SIGHUP, reloading checker configuration
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Initializing ipvs 2.6
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Netlink reflector reports IP 10.25.88.1 added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Netlink reflector reports IP 10.25.88.169 added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Netlink reflector reports IP 10.25.88.166 added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Netlink reflector reports IP 10.25.88.27 added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Netlink reflector reports IP 10.25.88.87 added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Netlink reflector reports IP 10.25.88.22 added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Netlink reflector reports IP 10.25.88.48 added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Netlink reflector reports IP fe80::21b:21ff:fecc:1c1c added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Netlink reflector reports IP fe80::21b:21ff:fecc:1d3d added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Netlink reflector reports IP fe80::21b:21ff:fecc:1d3c added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Netlink reflector reports IP fe80::21b:21ff:fecc:1d3d added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Netlink reflector reports IP fe80::21b:21ff:fecc:1d3d added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Registering Kernel netlink reflector
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Netlink: Cannot open netlink socket : (Too many open files)
Jun 24 18:57:08 cnsz081823 Keepalived_vrrp[4635]: Netlink reflector reports IP 10.25.88.1 added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Error while registering Kernel netlink cmd channel
Jun 24 18:57:08 cnsz081823 Keepalived_vrrp[4635]: Netlink reflector reports IP 10.25.88.169 added
Jun 24 18:57:08 cnsz081823 Keepalived_vrrp[4635]: Netlink reflector reports IP 10.25.88.166 added
Jun 24 18:57:08 cnsz081823 Keepalived_vrrp[4635]: Netlink reflector reports IP 10.25.88.27 added
Jun 24 18:57:08 cnsz081823 Keepalived_vrrp[4635]: Netlink reflector reports IP 10.25.88.87 added
Jun 24 18:57:08 cnsz081823 Keepalived_vrrp[4635]: Netlink reflector reports IP 10.25.88.22 added
Jun 24 18:57:08 cnsz081823 Keepalived_vrrp[4635]: Netlink reflector reports IP 10.25.88.48 added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Opening file '/etc/keepalived/keepalived.conf'.
Jun 24 18:57:08 cnsz081823 Keepalived_vrrp[4635]: Netlink reflector reports IP fe80::21b:21ff:fecc:1c1c added
Jun 24 18:57:08 cnsz081823 Keepalived_healthcheckers[4634]: Configuration file '/etc/keepalived/keepalived.conf' open problem (Too many open files)...
Jun 24 18:57:08 cnsz081823 Keepalived_vrrp[4635]: Netlink reflector reports IP fe80::21b:21ff:fecc:1d3d added
Jun 24 18:57:08 cnsz081823 Keepalived_vrrp[4635]: Netlink reflector reports IP fe80::21b:21ff:fecc:1d3c added
......
```

**Comment by pqarmitage** [comment]
I can confirm that there is a problem with keepalived v1.2.13 whereby the keepalived_healthchecker process leaks a file descriptor on configuration reload. The latest version of keepalived available as a Centos RPM for Centos 7 is v1.3.5, and this does not have the same problem. I recommend you update to resolve the problem.

Better still, build a current version of keepalived, e.g. v2.0.16 and use that; there have been hundreds of fixes and enhancements since v1.2.13.
